File: backend/internal/stlparser/stl_parser.go
package stlparser

import (
	"bytes"
	"encoding/binary"
	"errors"
	"fmt"
	"io"
	"math"
	"strings"
)

type Point3D struct {
	X float64 `json:"x"`
	Y float64 `json:"y"`
	Z float64 `json:"z"`
}

type Triangle struct {
	Normal   Point3D `json:"normal"`
	V1       Point3D `json:"v1"`
	V2       Point3D `json:"v2"`
	V3       Point3D `json:"v3"`
}

type MeshAnalysis struct {
	TriangleCount  int     `json:"triangle_count"`
	VolumeCm3      float64 `json:"volume_cm3"`
	SurfaceAreaCm2 float64 `json:"surface_area_cm2"`
	DimXMm         float64 `json:"dim_x_mm"`
	DimYMm         float64 `json:"dim_y_mm"`
	DimZMm         float64 `json:"dim_z_mm"`
}

type SliceResult struct {
	Mesh              MeshAnalysis `json:"mesh"`
	MaterialType      string       `json:"material_type"`
	InfillPercent     int          `json:"infill_percent"`
	EstimatedWeightG  float64      `json:"estimated_weight_g"`
	EstimatedHours    float64      `json:"estimated_hours"`
	EstimatedPrice    float64      `json:"estimated_price"`
}

// Material densities in g/cm3
var materialDensities = map[string]float64{
	"PLA":   1.24,
	"ABS":   1.04,
	"PETG":  1.27,
	"TPU":   1.21,
	"Resin": 1.15,
}

// ParseSTL auto-detects binary or ASCII STL data and computes mesh volume, dimensions, and area
func ParseSTL(reader io.Reader) (MeshAnalysis, error) {
	data, err := io.ReadAll(reader)
	if err != nil {
		return MeshAnalysis{}, fmt.Errorf("erro ao ler arquivo STL: %w", err)
	}

	if len(data) < 84 {
		return MeshAnalysis{}, errors.New("tamanho de arquivo STL inválido ou corrompido")
	}

	// Check if ASCII STL format
	headerPrefix := strings.ToLower(string(data[:80]))
	if strings.HasPrefix(strings.TrimSpace(headerPrefix), "solid") && !isBinarySTL(data) {
		return parseASCIISTL(data)
	}

	return parseBinarySTL(data)
}

func isBinarySTL(data []byte) bool {
	if len(data) < 84 {
		return false
	}
	expectedNumTriangles := binary.LittleEndian.Uint32(data[80:84])
	expectedSize := 84 + (expectedNumTriangles * 50)
	return uint64(len(data)) == uint64(expectedSize)
}

func parseBinarySTL(data []byte) (MeshAnalysis, error) {
	if len(data) < 84 {
		return MeshAnalysis{}, errors.New("tamanho insuficiente para STL binário")
	}

	numTriangles := binary.LittleEndian.Uint32(data[80:84])
	if numTriangles == 0 {
		return MeshAnalysis{}, errors.New("arquivo STL não contém triângulos")
	}

	var totalSignedVolume float64
	var totalArea float64

	minX, maxX := math.MaxFloat64, -math.MaxFloat64
	minY, maxY := math.MaxFloat64, -math.MaxFloat64
	minZ, maxZ := math.MaxFloat64, -math.MaxFloat64

	offset := 84
	dataLen := len(data)

	for i := 0; i < int(numTriangles); i++ {
		if offset+50 > dataLen {
			break
		}

		v1 := readPoint(data[offset+12 : offset+24])
		v2 := readPoint(data[offset+24 : offset+36])
		v3 := readPoint(data[offset+36 : offset+48])
		offset += 50

		// Update bounding box
		minX = math.Min(minX, math.Min(v1.X, math.Min(v2.X, v3.X)))
		maxX = math.Max(maxX, math.Max(v1.X, math.Max(v2.X, v3.X)))

		minY = math.Min(minY, math.Min(v1.Y, math.Min(v2.Y, v3.Y)))
		maxY = math.Max(maxY, math.Max(v1.Y, math.Max(v2.Y, v3.Y)))

		minZ = math.Min(minZ, math.Min(v1.Z, math.Min(v2.Z, v3.Z)))
		maxZ = math.Max(maxZ, math.Max(v1.Z, math.Max(v2.Z, v3.Z)))

		// Signed tetrahedron volume calculation
		vSigned := (-v3.X*v2.Y*v1.Z + v2.X*v3.Y*v1.Z + v3.X*v1.Y*v2.Z - v1.X*v3.Y*v2.Z - v2.X*v1.Y*v3.Z + v1.X*v2.Y*v3.Z) / 6.0
		totalSignedVolume += vSigned

		// Surface area calculation
		area := calculateTriangleArea(v1, v2, v3)
		totalArea += area
	}

	volMm3 := math.Abs(totalSignedVolume)
	volCm3 := volMm3 / 1000.0
	areaCm2 := totalArea / 100.0

	dimX := math.Max(0, maxX-minX)
	dimY := math.Max(0, maxY-minY)
	dimZ := math.Max(0, maxZ-minZ)

	return MeshAnalysis{
		TriangleCount:  int(numTriangles),
		VolumeCm3:      math.Round(volCm3*100) / 100,
		SurfaceAreaCm2: math.Round(areaCm2*100) / 100,
		DimXMm:         math.Round(dimX*10) / 10,
		DimYMm:         math.Round(dimY*10) / 10,
		DimZMm:         math.Round(dimZ*10) / 10,
	}, nil
}

func parseASCIISTL(data []byte) (MeshAnalysis, error) {
	lines := strings.Split(string(data), "\n")
	var vertices []Point3D
	var totalSignedVolume float64
	var totalArea float64
	triCount := 0

	minX, maxX := math.MaxFloat64, -math.MaxFloat64
	minY, maxY := math.MaxFloat64, -math.MaxFloat64
	minZ, maxZ := math.MaxFloat64, -math.MaxFloat64

	for _, line := range lines {
		line = strings.TrimSpace(line)
		if strings.HasPrefix(line, "vertex ") {
			var p Point3D
			_, err := fmt.Sscanf(line, "vertex %f %f %f", &p.X, &p.Y, &p.Z)
			if err == nil {
				vertices = append(vertices, p)
				minX = math.Min(minX, p.X)
				maxX = math.Max(maxX, p.X)
				minY = math.Min(minY, p.Y)
				maxY = math.Max(maxY, p.Y)
				minZ = math.Min(minZ, p.Z)
				maxZ = math.Max(maxZ, p.Z)

				if len(vertices) == 3 {
					v1, v2, v3 := vertices[0], vertices[1], vertices[2]
					vSigned := (-v3.X*v2.Y*v1.Z + v2.X*v3.Y*v1.Z + v3.X*v1.Y*v2.Z - v1.X*v3.Y*v2.Z - v2.X*v1.Y*v3.Z + v1.X*v2.Y*v3.Z) / 6.0
					totalSignedVolume += vSigned
					totalArea += calculateTriangleArea(v1, v2, v3)
					triCount++
					vertices = vertices[:0]
				}
			}
		}
	}

	if triCount == 0 {
		return MeshAnalysis{}, errors.New("nenhum vértice válido encontrado no arquivo ASCII STL")
	}

	volCm3 := math.Abs(totalSignedVolume) / 1000.0
	areaCm2 := totalArea / 100.0

	return MeshAnalysis{
		TriangleCount:  triCount,
		VolumeCm3:      math.Round(volCm3*100) / 100,
		SurfaceAreaCm2: math.Round(areaCm2*100) / 100,
		DimXMm:         math.Round(math.Max(0, maxX-minX)*10) / 10,
		DimYMm:         math.Round(math.Max(0, maxY-minY)*10) / 10,
		DimZMm:         math.Round(math.Max(0, maxZ-minZ)*10) / 10,
	}, nil
}

func readPoint(b []byte) Point3D {
	return Point3D{
		X: float64(math.Float32frombits(binary.LittleEndian.Uint32(b[0:4]))),
		Y: float64(math.Float32frombits(binary.LittleEndian.Uint32(b[4:8]))),
		Z: float64(math.Float32frombits(binary.LittleEndian.Uint32(b[8:12]))),
	}
}

func calculateTriangleArea(p1, p2, p3 Point3D) float64 {
	ax, ay, az := p2.X-p1.X, p2.Y-p1.Y, p2.Z-p1.Z
	bx, by, bz := p3.X-p1.X, p3.Y-p1.Y, p3.Z-p1.Z

	cx := ay*bz - az*by
	cy := az*bx - ax*bz
	cz := ax*by - ay*bx

	return 0.5 * math.Sqrt(cx*cx+cy*cy+cz*cz)
}

// CalculateWeightAndHours estimates 3D print weight (g), hours (h), and suggested price
func CalculateWeightAndHours(mesh MeshAnalysis, material string, infillPercent int) SliceResult {
	if material == "" {
		material = "PLA"
	}
	density, ok := materialDensities[strings.ToUpper(material)]
	if !ok {
		density = 1.24
	}

	if infillPercent <= 0 {
		infillPercent = 20
	} else if infillPercent > 100 {
		infillPercent = 100
	}

	// Shell ratio: ~20% walls, top and bottom solid layers
	shellRatio := 0.20
	infillRatio := float64(infillPercent) / 100.0
	effectiveVolRatio := shellRatio + (1.0-shellRatio)*infillRatio

	effectiveVolCm3 := mesh.VolumeCm3 * effectiveVolRatio
	weightG := math.Max(5.0, math.Round(effectiveVolCm3*density*10)/10)

	// Extrusion print speed: ~16g per hour
	hours := math.Max(0.5, math.Round((weightG/16.0)*10)/10)

	// Pricing calculation: Material cost + Machine hour cost
	matMultiplier := 1.0
	switch strings.ToUpper(material) {
	case "RESIN":
		matMultiplier = 1.8
	case "PETG":
		matMultiplier = 1.25
	case "TPU":
		matMultiplier = 1.5
	case "ABS":
		matMultiplier = 1.15
	}

	price := math.Round((weightG*0.45 + hours*14.0) * matMultiplier)
	if price < 25.0 {
		price = 25.0
	}

	return SliceResult{
		Mesh:             mesh,
		MaterialType:     material,
		InfillPercent:    infillPercent,
		EstimatedWeightG: weightG,
		EstimatedHours:   hours,
		EstimatedPrice:   price,
	}
}

// Ensure bytes buffer implements io.Reader
var _ io.Reader = (*bytes.Reader)(nil)

File: frontend/src/utils/stlParser.ts
export interface ClientMeshAnalysis {
  triangleCount: number;
  volumeCm3: number;
  surfaceAreaCm2: number;
  dimXMm: number;
  dimYMm: number;
  dimZMm: number;
}

export interface ClientSliceResult {
  mesh: ClientMeshAnalysis;
  materialType: string;
  infillPercent: number;
  estimatedWeightG: number;
  estimatedHours: number;
  estimatedPrice: number;
}

const MATERIAL_DENSITIES: Record<string, number> = {
  PLA: 1.24,
  ABS: 1.04,
  PETG: 1.27,
  TPU: 1.21,
  Resin: 1.15,
};

/**
 * Fast client-side ArrayBuffer STL binary/ASCII parser
 */
export function parseSTLArrayBuffer(buffer: ArrayBuffer): ClientMeshAnalysis {
  const dataView = new DataView(buffer);
  const dataLen = buffer.byteLength;

  if (dataLen < 84) {
    throw new Error('Arquivo STL muito pequeno ou corrompido.');
  }

  // Check if Binary STL
  const numTriangles = dataView.getUint32(80, true);
  const expectedBinarySize = 84 + numTriangles * 50;

  if (dataLen === expectedBinarySize) {
    return parseBinarySTL(dataView, numTriangles);
  }

  // ASCII fallback
  const decoder = new TextDecoder('utf-8');
  const text = decoder.decode(buffer);
  return parseASCIISTL(text);
}

function parseBinarySTL(view: DataView, numTriangles: number): ClientMeshAnalysis {
  let totalSignedVolume = 0;
  let totalArea = 0;

  let minX = Infinity, maxX = -Infinity;
  let minY = Infinity, maxY = -Infinity;
  let minZ = Infinity, maxZ = -Infinity;

  let offset = 84;

  for (let i = 0; i < numTriangles; i++) {
    if (offset + 50 > view.byteLength) break;

    // Normal vector (12 bytes, offset + 0)
    // Vertex 1 (12 bytes, offset + 12)
    const v1x = view.getFloat32(offset + 12, true);
    const v1y = view.getFloat32(offset + 16, true);
    const v1z = view.getFloat32(offset + 20, true);

    // Vertex 2 (12 bytes, offset + 24)
    const v2x = view.getFloat32(offset + 24, true);
    const v2y = view.getFloat32(offset + 28, true);
    const v2z = view.getFloat32(offset + 32, true);

    // Vertex 3 (12 bytes, offset + 36)
    const v3x = view.getFloat32(offset + 36, true);
    const v3y = view.getFloat32(offset + 40, true);
    const v3z = view.getFloat32(offset + 44, true);

    offset += 50;

    minX = Math.min(minX, v1x, v2x, v3x);
    maxX = Math.max(maxX, v1x, v2x, v3x);

    minY = Math.min(minY, v1y, v2y, v3y);
    maxY = Math.max(maxY, v1y, v2y, v3y);

    minZ = Math.min(minZ, v1z, v2z, v3z);
    maxZ = Math.max(maxZ, v1z, v2z, v3z);

    // Signed volume of tetrahedron
    const vSigned = (-v3x * v2y * v1z + v2x * v3y * v1z + v3x * v1y * v2z - v1x * v3y * v2z - v2x * v1y * v3z + v1x * v2y * v3z) / 6.0;
    totalSignedVolume += vSigned;

    // Triangle Area
    const ax = v2x - v1x, ay = v2y - v1y, az = v2z - v1z;
    const bx = v3x - v1x, by = v3y - v1y, bz = v3z - v1z;
    const cx = ay * bz - az * by;
    const cy = az * bx - ax * bz;
    const cz = ax * by - ay * bx;
    totalArea += 0.5 * Math.sqrt(cx * cx + cy * cy + cz * cz);
  }

  const volCm3 = Math.abs(totalSignedVolume) / 1000.0;
  const areaCm2 = totalArea / 100.0;

  return {
    triangleCount: numTriangles,
    volumeCm3: Math.round(volCm3 * 100) / 100,
    surfaceAreaCm2: Math.round(areaCm2 * 100) / 100,
    dimXMm: Math.round(Math.max(0, maxX - minX) * 10) / 10,
    dimYMm: Math.round(Math.max(0, maxY - minY) * 10) / 10,
    dimZMm: Math.round(Math.max(0, maxZ - minZ) * 10) / 10,
  };
}

function parseASCIISTL(text: string): ClientMeshAnalysis {
  const lines = text.split('\n');
  const vertices: { x: number; y: number; z: number }[] = [];
  let totalSignedVolume = 0;
  let totalArea = 0;
  let triCount = 0;

  let minX = Infinity, maxX = -Infinity;
  let minY = Infinity, maxY = -Infinity;
  let minZ = Infinity, maxZ = -Infinity;

  for (let i = 0; i < lines.length; i++) {
    const line = lines[i].trim();
    if (line.startsWith('vertex ')) {
      const parts = line.split(/\s+/);
      if (parts.length >= 4) {
        const x = parseFloat(parts[1]);
        const y = parseFloat(parts[2]);
        const z = parseFloat(parts[3]);
        if (!isNaN(x) && !isNaN(y) && !isNaN(z)) {
          vertices.push({ x, y, z });
          minX = Math.min(minX, x);
          maxX = Math.max(maxX, x);
          minY = Math.min(minY, y);
          maxY = Math.max(maxY, y);
          minZ = Math.min(minZ, z);
          maxZ = Math.max(maxZ, z);

          if (vertices.length === 3) {
            const v1 = vertices[0], v2 = vertices[1], v3 = vertices[2];
            const vSigned = (-v3.x * v2.y * v1.z + v2.x * v3.y * v1.z + v3.x * v1.y * v2.z - v1.x * v3.y * v2.z - v2.x * v1.y * v3.z + v1.x * v2.y * v3.z) / 6.0;
            totalSignedVolume += vSigned;

            const ax = v2.x - v1.x, ay = v2.y - v1.y, az = v2.z - v1.z;
            const bx = v3.x - v1.x, by = v3.y - v1.y, bz = v3.z - v1.z;
            const cx = ay * bz - az * by;
            const cy = az * bx - ax * bz;
            const cz = ax * by - ay * bx;
            totalArea += 0.5 * Math.sqrt(cx * cx + cy * cy + cz * cz);

            triCount++;
            vertices.length = 0;
          }
        }
      }
    }
  }

  const volCm3 = Math.abs(totalSignedVolume) / 1000.0;
  const areaCm2 = totalArea / 100.0;

  return {
    triangleCount: triCount,
    volumeCm3: Math.round(volCm3 * 100) / 100,
    surfaceAreaCm2: Math.round(areaCm2 * 100) / 100,
    dimXMm: Math.round(Math.max(0, maxX - minX) * 10) / 10,
    dimYMm: Math.round(Math.max(0, maxY - minY) * 10) / 10,
    dimZMm: Math.round(Math.max(0, maxZ - minZ) * 10) / 10,
  };
}

export function calculateClientSlice(mesh: ClientMeshAnalysis, material = 'PLA', infillPercent = 20): ClientSliceResult {
  const density = MATERIAL_DENSITIES[material] || 1.24;
  const shellRatio = 0.20;
  const infillRatio = Math.min(100, Math.max(0, infillPercent)) / 100.0;
  const effectiveVolRatio = shellRatio + (1.0 - shellRatio) * infillRatio;

  const effectiveVolCm3 = mesh.volumeCm3 * effectiveVolRatio;
  const weightG = Math.max(5.0, Math.round(effectiveVolCm3 * density * 10) / 10);
  const hours = Math.max(0.5, Math.round((weightG / 16.0) * 10) / 10);

  let matMultiplier = 1.0;
  if (material === 'Resin') matMultiplier = 1.8;
  else if (material === 'PETG') matMultiplier = 1.25;
  else if (material === 'TPU') matMultiplier = 1.5;
  else if (material === 'ABS') matMultiplier = 1.15;

  let price = Math.round((weightG * 0.45 + hours * 14.0) * matMultiplier);
  if (price < 25.0) price = 25.0;

  return {
    mesh,
    materialType: material,
    infillPercent,
    estimatedWeightG: weightG,
    estimatedHours: hours,
    estimatedPrice: price,
  };
}
